File: src/data/scripts/campaign/AI/niko_MPC_satelliteFleetAI.java
package data.scripts.campaign.AI;

import com.fs.starfarer.api.campaign.BattleAPI;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.FleetEncounterContextPlugin;
import com.fs.starfarer.campaign.ai.ModularFleetAI;
import com.fs.starfarer.campaign.fleet.CampaignFleet;
import data.scripts.campaign.misc.niko_MPC_satelliteHandler;
import data.utilities.niko_MPC_debugUtils;
import data.utilities.niko_MPC_satelliteBattleTracker;
import data.utilities.niko_MPC_satelliteUtils;

public class niko_MPC_satelliteFleetAI extends ModularFleetAI {

    public niko_MPC_satelliteFleetAI(CampaignFleet campaignFleet) {
        super(campaignFleet);
    }

    @Override
    public boolean wantsToJoin(BattleAPI battle, boolean considerPlayTransponderStatus) {
        if (!niko_MPC_debugUtils.assertEntityHasSatellites(getFleet())) return true;

        niko_MPC_satelliteHandler handler = niko_MPC_satelliteUtils.getSatelliteHandler(getFleet());
        niko_MPC_satelliteBattleTracker tracker = niko_MPC_satelliteUtils.getSatelliteBattleTracker();

        if (tracker.areSatellitesInvolvedInBattle(battle, handler)) {
            return false;
        }

        return true;
    }

   @Override
   public EncounterOption pickEncounterOption(FleetEncounterContextPlugin context, CampaignFleetAPI otherFleet, boolean pureCheck) {

        CampaignFleetAPI satelliteFleet = getFleet(); //todo: this fucking sucks
        BattleAPI battle = satelliteFleet.getBattle();
        if (battle != null) {
            float effectiveHostileStrength = 0;
            for (CampaignFleetAPI hostileFleet : battle.getOtherSideFor(satelliteFleet)) {
                effectiveHostileStrength += hostileFleet.getEffectiveStrength();
            }
            if (satelliteFleet.getEffectiveStrength() < (effectiveHostileStrength)) {
                return EncounterOption.HOLD_VS_STRONGER;
            }
            else return EncounterOption.HOLD;
        }
        else {
            if ((satelliteFleet.getEffectiveStrength() < (otherFleet.getEffectiveStrength()))) {
                return EncounterOption.HOLD_VS_STRONGER;
            }
        }
        return EncounterOption.HOLD;
    }

    @Override
    public EncounterOption pickEncounterOption(FleetEncounterContextPlugin context, CampaignFleetAPI otherFleet) {
        return pickEncounterOption(context, otherFleet, false);
    }
}

File: src/data/scripts/campaign/AI/niko_MPC_satelliteFleetAI.kt
package data.scripts.campaign.AI

import com.fs.starfarer.api.campaign.BattleAPI
import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.FleetEncounterContextPlugin
import com.fs.starfarer.api.campaign.ai.CampaignFleetAIAPI.EncounterOption
import com.fs.starfarer.campaign.ai.ModularFleetAI
import com.fs.starfarer.campaign.fleet.CampaignFleet
import data.scripts.campaign.econ.conditions.defenseSatellite.handlers.niko_MPC_satelliteHandlerCore
import data.utilities.niko_MPC_fleetUtils.getSatelliteEntityHandler
import data.utilities.niko_MPC_fleetUtils.isDummyFleet
import data.utilities.niko_MPC_satelliteUtils.getSatelliteBattleTracker

class niko_MPC_satelliteFleetAI(campaignFleet: CampaignFleet?) : ModularFleetAI(campaignFleet) {
    override fun wantsToJoin(battle: BattleAPI?, considerPlayTransponderStatus: Boolean): Boolean {
        val handler: niko_MPC_satelliteHandlerCore = fleet.getSatelliteEntityHandler() ?: return true
        val tracker = getSatelliteBattleTracker() ?: return false
        if (battle == null) return false
        return tracker.areSatellitesInvolvedInBattle(battle, handler)
    }

    override fun pickEncounterOption(
        context: FleetEncounterContextPlugin?,
        otherFleet: CampaignFleetAPI?,
        pureCheck: Boolean
    ): EncounterOption {
        val satelliteFleet: CampaignFleetAPI = fleet //todo: this fucking sucks
        val battle = satelliteFleet.battle
        if (battle != null) {
            var effectiveHostileStrength = 0f
            for (hostileFleet in battle.getOtherSideFor(satelliteFleet)) {
                effectiveHostileStrength += hostileFleet.effectiveStrength
            }
            return if (satelliteFleet.effectiveStrength < effectiveHostileStrength) {
                EncounterOption.HOLD_VS_STRONGER
            } else EncounterOption.HOLD
        } else if (otherFleet != null){
            if (satelliteFleet.effectiveStrength < otherFleet.effectiveStrength) {
                return EncounterOption.HOLD_VS_STRONGER
            }
        }
        return EncounterOption.HOLD
    }

    override fun pickEncounterOption(
        context: FleetEncounterContextPlugin?,
        otherFleet: CampaignFleetAPI?
    ): EncounterOption {
        return pickEncounterOption(context, otherFleet, false)
    }
}